Added documentation for Salesforce authentication password plugin.

weissleb opened the pull request:
@dain here is the documentation for the Salesforce Authenticator, #4372 .  Please advise if changes are needed.

weissleb commented:
@lhofhansl please feel free to review these docs as well.  Not required, but if you have time and the inclination.

mosabua commented on presto-docs/src/main/sphinx/security/salesforce.rst:
```diff
@@ -0,0 +1,78 @@
+============================
+
+Presto can be configured to enable frontend password authentication over
+HTTPS for clients, such as the CLI, or the JDBC and ODBC drivers. The
```
What about the web ui?

dain replied:
This is a password authenticater and will just work with the form based authentication in the Web UI.  The other authenticater don't really mention the web UI either.

weissleb replied:
Yup, I've tested this previously, https://prestosql.io/docs/current/admin/web-interface.html#authentication.

mosabua replied:
So maybe we should call this out as well then

mosabua commented on presto-docs/src/main/sphinx/security/salesforce.rst:
```diff
@@ -0,0 +1,78 @@
+username and password (or password and :ref:`security token` concatenation)
+are validated by having the Presto coordinator perform a login to Salesforce.
+
+This allows Presto admins to enable users to authenticate to Presto via their Salesforce
```
This allows you ..

weissleb replied:
Will do.

mosabua commented on presto-docs/src/main/sphinx/security/salesforce.rst:
```diff
@@ -0,0 +1,78 @@
+are validated by having the Presto coordinator perform a login to Salesforce.
+
+This allows Presto admins to enable users to authenticate to Presto via their Salesforce
+basic credentials. This is *not* a Salesforce connector, and does not allow users to query
```
Pull that sentence out of this paragraph  and into a separate note after it 

```
.. note:

  This is ... 
```

weissleb replied:
Will do.

mosabua commented on presto-docs/src/main/sphinx/security/salesforce.rst:
```diff
@@ -0,0 +1,78 @@
+
+This allows Presto admins to enable users to authenticate to Presto via their Salesforce
+basic credentials. This is *not* a Salesforce connector, and does not allow users to query
+Salesforce data. It's simply a means by which users can authenticate to Presto, similar
```
It enables users to authenticate to Presto, similar...

weissleb replied:
@mosabua , sorry I'm not quite sure I'm following this one.  I'm changing line 10, per your first comment.

mosabua commented on presto-docs/src/main/sphinx/security/salesforce.rst:
```diff
@@ -0,0 +1,78 @@
+--------------------------------------
+
+Enable Salesforce authentication by creating an
+``etc/password-authenticator.properties`` file on the coordinator:
```
Only required on the coordinator? Optional to be on workers?

weissleb replied:
I believe it's only on the coordinator, and not necessary on the workers (though it doesn't hurt anything).  I pretty much copied this sentence from https://prestosql.io/docs/current/security/password-file.html#password-authenticator-configuration.

mosabua replied:
ok

mosabua commented on presto-docs/src/main/sphinx/security/salesforce.rst:
```diff
@@ -0,0 +1,78 @@
+Enable Salesforce authentication by creating an
+``etc/password-authenticator.properties`` file on the coordinator:
+
+.. code-block:: none
```
use properties

weissleb replied:
Will do.

mosabua commented on presto-docs/src/main/sphinx/security/salesforce.rst:
```diff
@@ -0,0 +1,78 @@
+
+The following configuration properties are available:
+
+====================================   ==============================================
```
This could (maybe should) be formatted as a list-table or a definition list.

mosabua replied:
E.g. see https://github.com/prestosql/presto/blob/master/presto-docs/src/main/sphinx/connector/hive-caching.rst

weissleb replied:
@mosabua I copied the table at https://prestosql.io/docs/current/security/password-file.html#password-authenticator-configuration.  Happy to change it if you'd prefer.  Please advise.

mosabua replied:
lets leave it for now I guess ..

mosabua commented on presto-docs/src/main/sphinx/security/salesforce.rst:
```diff
@@ -0,0 +1,78 @@
+Property                               Description
+====================================   ==============================================
+``salesforce.allowed-organizations``   Comma separated list of 18 character Salesforce.com
+                                       Organization ID's for a second simple layer of security.
```
organization

mosabua replied:
second, simple layer

weissleb replied:
Will do.

weissleb replied:
I capitalize Organization within the context of Salesforce.com Organization ID because that's how it appears in the Salesforce UI. Hopefully this is ok.

mosabua replied:
ok .. seems fine to me then

mosabua commented on presto-docs/src/main/sphinx/security/salesforce.rst:
```diff
@@ -0,0 +1,78 @@
+``salesforce.allowed-organizations``   Comma separated list of 18 character Salesforce.com
+                                       Organization ID's for a second simple layer of security.
+                                       This option can be explicitly ignored using ``all``
+                                       here, which will bypass any check of the authenticated
```
ignored using ``all``, which bypasses any check

weissleb replied:
Will do.

mosabua commented on presto-docs/src/main/sphinx/security/salesforce.rst:
```diff
@@ -0,0 +1,78 @@
+                                       here, which will bypass any check of the authenticated
+                                       user's Salesforce Organization Id.
+
+``salesforce.cache-size``              Maximum size of the cache that holds authenticated users.
```
whats the unit here? number of users or kb or ?

weissleb replied:
Sorry, changing to `Max number of cached authenticated users.`.

mosabua commented on presto-docs/src/main/sphinx/security/salesforce.rst:
```diff
@@ -0,0 +1,78 @@
+``salesforce.cache-size``              Maximum size of the cache that holds authenticated users.
+                                       Defaults to ``4096``.
+
+``salesforce.cache-expire-duration``   Expire duration for an entry in cache since last write.
```
whats the unit here?

mosabua replied:
Also reword description ... maybe something like

```
Duration for an entry to remain the cache after the last write before it is removed.
```

weissleb replied:
Will update the phrase.  

The type is `io.airlift.units.Duration`, and the unit is in the form of [this pattern](https://github.com/airlift/units/blob/master/src/main/java/io/airlift/units/Duration.java#L39), where the unit string should match one of [these cases](https://github.com/airlift/units/blob/master/src/main/java/io/airlift/units/Duration.java#L215).  It's similar to [`file.refresh-period`](https://prestosql.io/docs/current/security/password-file.html#password-authenticator-configuration)

mosabua replied:
okay .. we probably should document that somewhere separately

mosabua commented on presto-docs/src/main/sphinx/security/salesforce.rst:
```diff
@@ -0,0 +1,78 @@
+
+Salesforce Concepts
+-------------------
+
```
intro sentence missing

weissleb replied:
Will add.

mosabua commented on presto-docs/src/main/sphinx/security/salesforce.rst:
```diff
@@ -0,0 +1,78 @@
+^^^^^^^^^^^^^^
+
+Credentials are a user's Salesforce username and password if Presto is connecting from a whitelisted
+IP, or username and password / security token concatenation otherwise.  For example, if the coordinator
```
No space around /

weissleb replied:
Will do.

mosabua commented on presto-docs/src/main/sphinx/security/salesforce.rst:
```diff
@@ -0,0 +1,78 @@
+^^^^^^^^^^^^^^
+
+Credentials are a user's Salesforce username and password if Presto is connecting from a whitelisted
+IP, or username and password / security token concatenation otherwise.  For example, if the coordinator
```
Should we talk about the coordinator here or just "Presto" since it could also be a DNS name pointing to a loadbalancer pointing to the coordinator ..

weissleb replied:
Good call.  Will change to just "Presto".

mosabua commented on presto-docs/src/main/sphinx/security/salesforce.rst:
```diff
@@ -0,0 +1,78 @@
+Credentials are a user's Salesforce username and password if Presto is connecting from a whitelisted
+IP, or username and password / security token concatenation otherwise.  For example, if the coordinator
+is not whitelisted, and your password is ``password`` and security token is ``token``, use ``passwordtoken``
+as your Presto password.
```
remove Presto

mosabua commented on presto-docs/src/main/sphinx/security/salesforce.rst:
```diff
@@ -0,0 +1,78 @@
+Credentials are a user's Salesforce username and password if Presto is connecting from a whitelisted
+IP, or username and password / security token concatenation otherwise.  For example, if the coordinator
+is not whitelisted, and your password is ``password`` and security token is ``token``, use ``passwordtoken``
+as your Presto password.
```
use ``passwordtoken`` to authenticate.

weissleb replied:
Will update.

mosabua commented on presto-docs/src/main/sphinx/security/salesforce.rst:
```diff
@@ -0,0 +1,78 @@
+is not whitelisted, and your password is ``password`` and security token is ``token``, use ``passwordtoken``
+as your Presto password.
+
+To learn more about the Salesforce security token, see the `reset security token documentation
```
You could just make the first mention of the security token in the first sentence a link to these docs and remove the whole paragraph here

weissleb replied:
Sounds good.

mosabua commented on presto-docs/src/main/sphinx/security/salesforce.rst:
```diff
@@ -0,0 +1,78 @@
+To learn more about the Salesforce security token, see the `reset security token documentation
+<https://help.salesforce.com/articleView?id=user_security_token.htm&type=5>`_.
+
+To learn more about whitelisting an IP range, see the `set trusted IP ranges documentation
```
Use descriptive link here as well. 

Eg.

```
You can configure a public IP for Presto as a trusted IP number by `whitelisting an IP range <url>`_.
```

weissleb replied:
Sounds good.

mosabua commented on presto-docs/src/main/sphinx/security/salesforce.rst:
```diff
@@ -0,0 +1,78 @@
+Salesforce Organization Id's
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+There is a secondary layer of security beyond simply providing credentials, which is that
```
You can configure a one or more 18 character Salesforce.com organization IDs. They can be used as a secondray layer of security.

weissleb replied:
Will update.

mosabua commented on presto-docs/src/main/sphinx/security/salesforce.rst:
```diff
@@ -0,0 +1,78 @@
+There is a secondary layer of security beyond simply providing credentials, which is that
+the admin can configure one or more 18 character Salesforce.com Organization ID's to which the
+user must belong.  When the user authenticates, the Salesforce API returns the 18 character
+Salesforce.com Organization ID for the user.  Presto's Salesforce authenticator will ensure this
```
organization

mosabua replied:
The Presto Salesforce authenticator ensures that the ID matches on of the IDs configured in ..

weissleb replied:
Will update the sentence.  I capitalize Organization within the context of Salesforce.com Organization ID because that's how it appears in the Salesforce UI.  Hopefully this is ok.

mosabua commented on presto-docs/src/main/sphinx/security/salesforce.rst:
```diff
@@ -0,0 +1,78 @@
+Salesforce.com Organization ID for the user.  Presto's Salesforce authenticator will ensure this
+matches one of the configured ``salesforce.allowed-organizations``.
+
+Optionally, the admin can configure ``all`` to explicitly ignore this layer of security.
```
Optionally you can configure

weissleb replied:
Will update.

mosabua commented on presto-docs/src/main/sphinx/security/salesforce.rst:
```diff
@@ -0,0 +1,78 @@
+
+Optionally, the admin can configure ``all`` to explicitly ignore this layer of security.
+
+Admins can find their Salesforce.com Organization ID by following the procedures `described here
```
Refactor so this paragraph uses descriptive links please. See https://developers.google.com/style/link-text

weissleb replied:
Gotcha.  Will do.

weissleb commented:
> Please address the requested changes and I can review again after.

@mosabua I've addressed all your comments.  I guess amending the previous commit wasn't the best idea.  Sorry about that.  Luckily, the changes are small and only to one file.

mosabua commented on presto-docs/src/main/sphinx/security/salesforce.rst:
```diff
@@ -0,0 +1,79 @@
+basic credentials.
+
+.. note::
+    This is *not* a Salesforce connector, and does not allow users to query
```
I think this requires a new line below note

mosabua commented on presto-docs/src/main/sphinx/security/salesforce.rst:
```diff
@@ -0,0 +1,79 @@
+
+Salesforce Concepts
+-------------------
+There are two Salesforce specific aspects to this authenticator.  They are the use of the
```
empty line above paragraph

mosabua commented on presto-docs/src/main/sphinx/security/salesforce.rst:
```diff
@@ -0,0 +1,79 @@
+Salesforce.com Organization IDs
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+You can configure a one or more 18 character Salesforce.com Organization IDs. They can be used as a
```
a one or more .. something is weird here

mosabua replied:
You can configure one or more Salesforce Organization IDs for additional authentication. The IDs need to be 18 character long.

weissleb commented:
> Only minor changes left.. imho this is ready to go.

Thanks @mosabua , I'll make those changes and issue another commit.

weissleb commented:
@electrum @dain sorry for the nudge here guys, if you get a moment, I'd appreciate your review on these docs.  Looks like @mosabua is a thumbs up.

electrum commented on presto-docs/src/main/sphinx/security/salesforce.rst:
```diff
@@ -0,0 +1,80 @@
+.. code-block:: properties
+
+    password-authenticator.name=salesforce
+    salesforce.allowed-organizations=allowed salesforce org id's, or the word all
```
How about
```
salesforce.allowed-organizations=<allowed-org-ids or all>

weissleb replied:
Will do.

electrum commented on presto-docs/src/main/sphinx/security/salesforce.rst:
```diff
@@ -0,0 +1,80 @@
+============================
```
Make these bars the same length as the text

weissleb replied:
Gotcha.

electrum commented on presto-docs/src/main/sphinx/security/salesforce.rst:
```diff
@@ -0,0 +1,80 @@
+
+Presto can be configured to enable frontend password authentication over
+HTTPS for clients, such as the CLI, or the JDBC and ODBC drivers. The
+username and password (or password and :ref:`security tokens` concatenation)
```
See explanation below
```
username and password (or password and `security tokens <#security-tokens>`__ concatenation)

weissleb replied:
Cool.  Honestly, I wanted to do this, but couldn't quite figure out how.  Thanks.

electrum commented on presto-docs/src/main/sphinx/security/salesforce.rst:
```diff
@@ -0,0 +1,80 @@
+There are two Salesforce specific aspects to this authenticator.  They are the use of the
+Salesforce security token, and configuration of one or more Salesforce.com Organization IDs.
+
+.. _security tokens:
```
This creates a global reference which is problematic since "security tokens" is rather generic. Instead, we can link directly to this section since it's inside the same document, so no need for a reference at all. I gave the syntax above.

weissleb replied:
Removing.

electrum commented on presto-docs/src/main/sphinx/security/salesforce.rst:
```diff
@@ -0,0 +1,80 @@
+
+====================================   ==============================================
+Property                               Description
+====================================   ==============================================
```
Make the bars as wide as the text

weissleb replied:
Aight.

electrum commented on presto-docs/src/main/sphinx/security/salesforce.rst:
```diff
@@ -0,0 +1,80 @@
+                                       bypasses any check of the authenticated user's
+                                       Salesforce.com Organization ID.
+
+``salesforce.cache-size``              Max number of cached authenticated users.
```
Max -> Maximum

weissleb replied:
Will do.

electrum commented on presto-docs/src/main/sphinx/security/salesforce.rst:
```diff
@@ -0,0 +1,80 @@
+``salesforce.cache-size``              Max number of cached authenticated users.
+                                       Defaults to ``4096``.
+
+``salesforce.cache-expire-duration``   Duration for an entry to remain in the cache after the
```
How about
```
How long a cached authentication should be considered valid.

weissleb replied:
Will do.

weissleb commented:
Addressed @electrum 's review and squashed into single commit.

dain commented:
@weissleb I think you missed the comment about a period on the commit message

weissleb commented:
> @weissleb I think you missed the comment about a period on the commit message

Yeah, sorry.  I just fixed the mssage.
